### DISCUSSION TITLE
How to predefine possible values?

### DISCUSSION THREAD

**Discussion by theDanielJLewis** [typegoose/typegoose#951] [Q&A]
(I tried searching couldn't find an answer in GitHub or the documentation, as far as I understand it.)

Is it possible to predefine possible values when I'm defining a prop?

For example:

```js
export class WidgetClass {
  @prop()
  public day: 'Sunday' | 'Monday' | 'Tuesday';
}
```

The goal is that the field (in this case, `day`) can only ever contain one of those values. So trying to save it with any other value will give me a TypeScript error.

Thanks in advance!

**Comment by hasezoey** [comment]
You can use [enums](https://typegoose.github.io/typegoose/docs/api/decorators/prop/#enum) (strings and numbers only), for anything else, you will have to do [manual validation](https://typegoose.github.io/typegoose/docs/api/decorators/prop/#validate).
Enums are checked at compile time via the enum types and at runtime validated via the `enum` option.

**Reply by theDanielJLewis** [reply]
Thanks! I need to learn more about enums because they never really made sense to me. Like they seem redundant for `VALUE = 'value'`. Or like a way to map values to numbers.

Yes, I clearly don't understand enums well enough.

**Reply by hasezoey** [reply]
>  I need to learn more about enums because they never really made sense to me. Like they seem redundant for VALUE = 'value'. Or like a way to map values to numbers

TL;DR: they are fancy limited-key objects with extra types support.

in typescript they are by default numbers (starting from `0` and adding plus one going downward in definition order), you can assign specific numbers if you want to be explicit (or dont want the auto numbering, also if you manually assign a value to like `100` then the next auto-value is `101`), otherwise it also supports assigning strings, where you will need to be explicit with every value (otherwise the enum will be `string | number`). (they also support computed values)

and because enums dont exist in JS natively, typescript emulates it by basically being a object with key and value (and reverse as well)

[Read more in the typescript documentation](https://www.typescriptlang.org/docs/handbook/enums.html).

---

as for what `enum`s in mongoose are, they are basically just a array of possible values (typegoose automatically translates typescript enums), similar to `validate: (value) => enumArray.includes(value)`.
(also note that typegoose only accepts one-type enums only, only `string` or `number`, not `string | number`)
